File: backend/ai.py
import re
import sys
from pathlib import Path
import ollama

class Config:
    SCRIPT_PATH = Path(__file__).parent.absolute()
    PROJECT_ROOT = SCRIPT_PATH.parent if "backend" in SCRIPT_PATH.name else SCRIPT_PATH
    MODEL_NAME = 'qwen2.5-coder:7b'

    IGNORED_DIRS = {'.venv', 'venv', '__pycache__', '.git', 'node_modules'}
    VALID_EXTENSIONS = {'.js', '.html', '.css', '.py', '.rs'}


class FileManager:

    @staticmethod
    def normalize_text(text: str) -> str:
        return re.sub(r'\s+', ' ', text).strip()

    @classmethod
    def load_context(cls) -> str:
        context =[]
        for file_path in Config.PROJECT_ROOT.rglob('*'):
            if cls._should_ignore(file_path):
                continue

            if file_path.is_file() and file_path.suffix in Config.VALID_EXTENSIONS:
                try:
                    content = file_path.read_text(encoding='utf-8')
                    rel_path = file_path.relative_to(Config.PROJECT_ROOT)
                    context.append(f"--- FILE: {rel_path} ---\n{content}")
                except Exception:
                    pass
        return "\n".join(context)

    @classmethod
    def apply_patches(cls, ai_response: str) -> None:

        pattern = r'<patch path=["\'](.*?)["\']>\s*<search>(.*?)</search>\s*<replace>(.*?)</replace>\s*</patch>'
        matches = re.findall(pattern, ai_response, re.DOTALL | re.IGNORECASE)

        if not matches:
            print("[!] HIBA: Az AI nem küldött érvényes javítócsomagot (XML patch hiányzik).")
            return

        for path_str, search_text, replace_text in matches:
            cls._apply_single_patch(path_str, search_text, replace_text)

    @classmethod
    def _apply_single_patch(cls, path_str: str, search_text: str, replace_text: str) -> None:
        target_path = Config.PROJECT_ROOT / path_str.strip().replace("\\", "/")

        if not target_path.exists():
            print(f"[!] HIBA: Fájl nem található: {path_str}")
            return

        original_content = target_path.read_text(encoding='utf-8')
        s_text = search_text.strip()
        r_text = replace_text.strip()

        if cls.normalize_text(s_text) in cls.normalize_text(original_content):
            new_content = original_content.replace(s_text, r_text)
            target_path.write_text(new_content, encoding='utf-8')
            print(f"[+] SIKER: {path_str}")
        else:
            print(f"[!] HIBA: Nem találom a módosítandó részt itt: {path_str}")

    @staticmethod
    def _should_ignore(file_path: Path) -> bool:
        in_ignored_dir = any(ignored in file_path.parts for ignored in Config.IGNORED_DIRS)
        is_self = file_path.name == Path(__file__).name
        return in_ignored_dir or is_self


class AICodeEditor:

    def __init__(self):
        self.model = Config.MODEL_NAME

    def _get_system_message(self) -> str:
        return (
            "You are a Senior Developer. Your only goal is to write correct, error-free code patches.\n"
            "Output ONLY XML patches. NO chat, NO explanations, NO markdown.\n\n"
            "CRITICAL RULE FOR PYTHON (.py files):\n"
            "Python is extremely sensitive to indentation. A single wrong space will cause an IndentationError and crash the server. "
            "Use 'elif' for subsequent conditions, do not nest 'if' statements incorrectly.\n\n"
            "EXAMPLE of adding a new operator to `calculator.py`:\n"
            "<patch path=\"backend_python/calculator.py\">\n"
            "<search>\n"
            "    if op == '/': return \"Hiba: 0-val osztás\" if b == 0 else a / b\n"
            "    return \"Ismeretlen művelet\"\n"
            "</search>\n"
            "<replace>\n"
            "    if op == '/': return \"Hiba: 0-val osztás\" if b == 0 else a / b\n"
            "    elif op == '%': return a % b\n"
            "    return \"Ismeretlen művelet\"\n"
            "</replace>\n"
            "</patch>\n\n"
            "Format: <patch path=\"file\"><search>old code</search><replace>new code</replace></patch>"
        )

    def process_task(self, user_request: str) -> None:
        codebase = FileManager.load_context()
        prompt = f"TASK: {user_request}\n\nCODEBASE:\n{codebase}"

        print(f"[*] AI dolgozik ({self.model})...")

        try:
            response = ollama.chat(
                model=self.model,
                messages=[
                    {'role': 'system', 'content': self._get_system_message()},
                    {'role': 'user', 'content': prompt}
                ],
                options={'temperature': 0}
            )
            FileManager.apply_patches(response['message']['content'])

        except Exception as e:
            print(f"[!!] Végzetes AI hiba: {e}")


def run_ai_task(user_request: str) -> None:
    editor = AICodeEditor()
    editor.process_task(user_request)

if __name__ == '__main__':
    if len(sys.argv) > 1:
        run_ai_task(sys.argv[1])
    else:
        print("Kérlek, adj meg egy feladatot argumentumként!")

File: backend_rust/src/ai.rs
use regex::Regex;
use std::fs;
use std::path::{Path, PathBuf};
use walkdir::WalkDir;

const MODEL: &str = "qwen2.5-coder:7b";

const IGNORED_DIRS: &[&str] = &[
    ".git",
    "venv",
    ".venv",
    "node_modules",
    "__pycache__",
    "target",
];

fn project_root() -> PathBuf {
    let current = std::env::current_dir().unwrap();

    if current.file_name().unwrap() == "backend_rust" {
        current.parent().unwrap().to_path_buf()
    } else {
        current
    }
}

fn is_ignored(path: &Path) -> bool {
    path.components().any(|c| {
        let s = c.as_os_str().to_string_lossy();
        IGNORED_DIRS.contains(&s.as_ref())
    })
}

fn load_context() -> String {
    let root = project_root();

    let extensions = ["rs", "py", "js", "html", "css"];

    let mut result = String::new();

    for entry in WalkDir::new(&root)
        .into_iter()
        .filter_map(Result::ok)
    {
        let path = entry.path();

        if is_ignored(path) {
            continue;
        }

        if !path.is_file() {
            continue;
        }

        let ext = match path.extension().and_then(|e| e.to_str()) {
            Some(e) => e,
            None => continue,
        };

        if !extensions.contains(&ext) {
            continue;
        }

        if let Ok(content) = fs::read_to_string(path) {
            if let Ok(rel) = path.strip_prefix(&root) {
                result.push_str(
                    &format!(
                        "--- FILE: {} ---\n{}\n",
                        rel.display(),
                        content
                    )
                );
            }
        }
    }

    result
}

fn apply_patches(ai_response: &str) {
    let root = get_project_root();
    let re = Regex::new(r"(?is)<patch path=[\x22\x27](.*?)[\x22\x27]>\s*<search>(.*?)</search>\s*<replace>(.*?)</replace>").unwrap();

    let mut found = false;
    for cap in re.captures_iter(ai_response) {
        found = true;
        let path_str = cap[1].trim();
        let search = cap[2].trim();
        let replace = cap[3].trim();

        let target_path = root.join(path_str);
        if target_path.exists() {
            if let Ok(content) = fs::read_to_string(&target_path) {
                if content.contains(search) {
                    let new_content = content.replace(search, replace);
                    fs::write(&target_path, new_content).unwrap();
                    println!("SIKER: {}", path_str);
                } else {
                    println!("HIBA: Nem található a keresett rész: {}", path_str);
                }
            }
        } else {
            println!("HIBA: A fájl nem létezik: {}", path_str);
        }
    }

    if !found {
        println!("Az AI nem küldött érvényes javítócsomagot.");
    }
}

    let cleaned = ai_response
        .replace("```xml", "")
        .replace("```", "");

    let regex = Regex::new(
        r#"(?s)<patch path=["'](.*?)["']>\s*<search>(.*?)</search>\s*<replace>(.*?)</replace>\s*</patch>"#
    ).unwrap();

    let root = project_root();

    let mut found = false;

    for cap in regex.captures_iter(&cleaned) {

        found = true;

        let path_str = cap[1].trim().replace("\\", "/");

        let search = cap[2].trim();

        let replace = cap[3].trim();

        let target = root.join(path_str);

        if !target.exists() {
            println!("Fájl nem található");
            continue;
        }

        let content = match fs::read_to_string(&target) {
            Ok(v) => v,
            Err(_) => continue,
        };

        if !content.contains(search) {
            println!("Nincs találat");
            continue;
        }

        let updated = content.replacen(search, replace, 1);

        if fs::write(&target, updated).is_ok() {
            println!("PATCH OK");
        }
    }

    if !found {
        println!("Nincs patch");
    }
}

pub async fn run_ai_task(task: String) {

    println!("AI dolgozik...");

    let codebase = load_context();

    let system_prompt = concat!(
        "You are a headless code editor. ",
        "Output ONLY XML patches. ",
        "No markdown. ",
        "No explanations. ",
        "Format: ",
        "<patch path=\"file\">",
        "<search>old</search>",
        "<replace>new</replace>",
        "</patch>"
    );

    let payload = serde_json::json!({
        "model": MODEL,
        "stream": false,
        "options": {
            "temperature": 0
        },
        "messages": [
            {
                "role": "system",
                "content": system_prompt
            },
            {
                "role": "user",
                "content": format!(
                    "TASK: {}\n\nCODEBASE:\n{}",
                    task,
                    codebase
                )
            }
        ]
    });

    let client = reqwest::Client::new();

    match client
        .post("http://127.0.0.1:11434/api/chat")
        .json(&payload)
        .send()
        .await
    {
        Ok(res) => {

            match res.json::<serde_json::Value>().await {

                Ok(json) => {

                    if let Some(content) =
                        json["message"]["content"].as_str()
                    {
                        apply_patches(content);
                    }
                }

                Err(e) => {
                    println!("JSON HIBA: {}", e);
                }
            }
        }

        Err(e) => {
            println!("REQWEST HIBA: {}", e);
        }
    }
}